Interface for creating default converters for binder API

mrminy opened the pull request:
Add ability to make default converters for any given datatypes and connect it to a Vaadin 8 binder. 

Partly solves the issue discussed here: https://stackoverflow.com/questions/43154002/does-vaadin-8-binderbindinstancefields-only-work-with-string-data-types

Also, it would remove a lot of boilerplate code for converting regular values as Integer and Double from String when using TextField. Similarly, this would make it easier to have forms using the old java.util.Date with the new Vaadin 8 Date[Time]Fields.

Fixes #9202 

This works more like a implementation suggestion. Would require some more work (tests and docs) if it were to be accepted.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/vaadin/framework/11306)
<!-- Reviewable:end -->

mrminy committed:
Update Binder.java

Added an optional default converter for automatically converting between classes when using {@link #bindInstanceFields(Object)}

mrminy committed:
Add files via upload

Added an optional converter interface that can be injected into the Vaadin 8 Binder. This makes it possible to setup default converters for your application.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/vaadin/framework?pullRequest=11306) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/vaadin/framework?pullRequest=11306) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/vaadin/framework?pullRequest=11306) it.</sub>

mrminy committed:
Update IDefaultConverter.java

mrminy committed:
Update IDefaultConverter.java

mrminy committed:
Update Binder.java

mrminy committed:
Update IDefaultConverter.java

mrminy committed:
Update IDefaultConverter.java

Commit:
Merge branch 'master' into master

anasmi commented:
Hi,

and thanks for submitting the pull request! 

Could you, please, run `mvn clean install` in the project? (By running default formatting will be applied)

Than we could go through the changes itself : )

mrminy committed:
Compiled with 'mvn clean install', moved IDefaultConverter to converter folder

mrminy committed:
Compiled with 'mvn clean install'

ZheSun88 commented:
Hi, 
Thanks for the contribution again. 
May I ask you to provide a small sample on the usage of this new feature? That would make it easier for the code review.  

Thanks.

Commit:
wip: refactor proposed default converters

Pushing for comments. Missing examples, tests and a setup for the
default configurator for an app.

Commit:
wip 2: revisited design

Still missing tests and example

ZheSun88 committed:
Merge branch 'master' into master

ZheSun88 commented:
Hi @mrminy , 

I saw you have closed this PR, may I know about your opinion? 

As Pekka had been working on this PR and tried to get the functionality into the framework a while ago. But according to the different priorities, he must focus on other projects. 

My apologies for not being active in this issue, again.
